DataFormats/Portable/interface/PortableHostCollectionReadRules.h

```diff
@@ -0,0 +1,83 @@
+#ifndef DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+#define DataFormats_Portable_interface_PortableHostCollectionReadRules_h
+
+#include <boost/core/demangle.hpp>
+
+// required by <RtypesImp.h>
+#define G__DICTIONARY
+
+#include <Rtypes.h>
+#include <RtypesImp.h>
+#include <TGenericClassInfo.h>
+#include <TVirtualObject.h>
+
+#include "DataFormats/Portable/interface/PortableHostCollection.h"
+#include "FWCore/Utilities/interface/concatenate.h"
+#include "FWCore/Utilities/interface/stringize.h"
+
+// read function for PortableHostCollection, called for every event
+template <typename T>
+static void readPortableHostCollection(char *target, TVirtualObject *from_buffer) {
+  // extract the actual types
+  using Collection = T;
+  using Layout = typename Collection::Layout;
+
+  // valid only for PortableHostCollection<T>
+  static_assert(std::is_same_v<Collection, PortableHostCollection<Layout>>);
+
+  // proxy for the object being read from file
+  struct OnFile {
+    Layout &layout_;
+  };
+
+  // address in memory of the buffer containing the object being read from file
+  char *address = static_cast<char *>(from_buffer->GetObject());
+  // offset of the "layout_" data member
+  static ptrdiff_t layout_offset = from_buffer->GetClass()->GetDataMemberOffset("layout_");
+  // reference to the Layout object being read from file
+  OnFile onfile = {*(Layout *)(address + layout_offset)};
```

Comment:
Just curious, what is the benefit of using the `OnFile` struct instead of assigning directly to `Layout&` here?

Reply:
There is no benefit in this case, it's just how the read rules generated by ROOT look like.

I've decided to adopt the same syntax so that the textual representation on line 63
```c++
    rule->fCode = type + "::ROOTReadStreamer(newObj, onfile.layout_)";
```
remains valid (even if it is not used).

Reply:
> Just curious, what is the benefit of using the OnFile struct instead of assigning directly to Layout& here?

This allows a clearer syntax to differentiate between the memory with data coming from the file and the memory corresponding to the live object (`newObj`).